### PR TITLE
Adaption to support EDSS

### PR DESCRIPTION
**Pull request by yunguowei**
We need to make some adaption to support EDSS on meta-secure-env layer, however, EDSS implamentation will be pulsar-binary-release layer.

### CODE REVIEW

File: classes/user-key-store.bbclass [hunk 1]
```diff
@@ -61,10 +62,12 @@ def mok_sb_keys_dir(d):
     return d.getVar('MOK_SB_KEYS_DIR', True) + '/'
 
 def sb_sign(input, output, d):
-    if d.getVar('MOK_SB', True) == '1':
-        mok_sb_sign(input, output, d)
-    elif d.getVar('UEFI_SB', True) == '1':
+    if uks_signing_model(d) == 'edss':
+        edss_sign_efi_image(input, output, d)
+    elif uks_signing_model(d) == 'sample':
         uefi_sb_sign(input, output, d)
```
Comment (jiazhang0) [inline classes/user-key-store.bbclass]: sb_sign() may be either signing for uefi secure boot or mok secure boot. You need to handle MOK_SB and UEFI_SB for user and sample signing models.

File: classes/user-key-store.bbclass [hunk 2]
```diff
@@ -61,10 +62,12 @@ def mok_sb_keys_dir(d):
     return d.getVar('MOK_SB_KEYS_DIR', True) + '/'
 
 def sb_sign(input, output, d):
-    if d.getVar('MOK_SB', True) == '1':
-        mok_sb_sign(input, output, d)
-    elif d.getVar('UEFI_SB', True) == '1':
+    if uks_signing_model(d) == 'edss':
+        edss_sign_efi_image(input, output, d)
```
Comment (jiazhang0) [inline classes/user-key-store.bbclass]: sb_sign() may be either signing for uefi secure boot or mok secure boot. You need to handle MOK_SB and UEFI_SB for user and sample signing models.

File: classes/user-key-store.bbclass [hunk 3]
```diff
@@ -61,10 +62,12 @@ def mok_sb_keys_dir(d):
...
+    elif uks_signing_model(d) == 'sample':
```
Comment (jiazhang0) [inline classes/user-key-store.bbclass]: sb_sign() may be either signing for uefi secure boot or mok secure boot. You need to handle MOK_SB and UEFI_SB for user and sample signing models.

File: recipes-bsp/shim/shim_git.bb [hunk 4]
```diff
@@ -62,53 +63,20 @@ EXTRA_OEMAKE = " \
...
-    except:
-        raise bb.build.FuncFailed('ERROR: Unable to create shim.p12')
+    shim_prepare_sb_keys(d)
 }
 addtask prepare_signing_keys after do_configure before do_compile
 
 python do_sign() {
-    uefi_sb_sign('${S}/shim${EFI_ARCH}.efi', '${B}/shim${EFI_ARCH}.efi.signed', d)
+    sb_sign('${S}/shim${EFI_ARCH}.efi', '${B}/shim${EFI_ARCH}.efi.signed', d)
+    sb_sign('${S}/mm${EFI_ARCH}.efi', '${B}/mm${EFI_ARCH}.efi.signed', d)
```
Comment (jiazhang0) [inline recipes-bsp/shim/shim_git.bb]: There is another .efi called fb.efi need to be signed.

File: classes/user-key-store.bbclass [hunk 5]
```diff
@@ -86,6 +86,53 @@ def mok_sb_sign(input, output, d):
...
+
+# Prepare signing keys for shim
+def shim_prepare_sb_keys(d):
+    if uks_signing_model(d) == 'sample':
+        shim_prepare_sb_sample_keys(d)
+    elif uks_signing_model(d) == 'edss':
+        shim_prepare_sb_edss_keys(d)
+    elif uks_signing_model(d) == 'user':
+        shim_prepare_sb_user_keys(d)
+
```
Comment (jiazhang0) [inline classes/user-key-store.bbclass]: Put edss case as the last one.

Comment (jiazhang0) [inline classes/user-key-store.bbclass]: In order of sample -> user -> edss

File: classes/user-key-store.bbclass [hunk 6]
```diff
@@ -86,6 +86,53 @@ def mok_sb_sign(input, output, d):
...
+    import shutil
+    shutil.copyfile(dir + 'shim_cert.pem', d.getVar('S', True) + '/shim.crt')
+    pem2der(dir + 'vendor_cert.pem', d.getVar('WORKDIR', True) + '/vendor_cert.cer', d)
+    d.appendVar('EXTRA_OEMAKE', ' VENDOR_CERT_FILE="${WORKDIR}/vendor_cert.cer"')
+
+def shim_prepare_sb_user_keys(d):
+    d.appendVar('EXTRA_OEMAKE', ' VENDOR_DBX_FILE="${WORKDIR}/vendor_dbx.esl"') 
+
+# Prepare signing keys for shim
+def shim_prepare_sb_keys(d):
```
Comment (jiazhang0) [inline classes/user-key-store.bbclass]: If MOK_SB is not defined, this function should not do anything.

File: classes/user-key-store.bbclass [hunk 7]
```diff
@@ -86,6 +86,53 @@ def mok_sb_sign(input, output, d):
...
+    # Prepare shim_cert(EDSS certificate) and vendor_cert
+    dir = mok_sb_keys_dir(d)
+
+    import shutil
+    shutil.copyfile(dir + 'shim_cert.pem', d.getVar('S', True) + '/shim.crt')
+    pem2der(dir + 'vendor_cert.pem', d.getVar('WORKDIR', True) + '/vendor_cert.cer', d)
+    d.appendVar('EXTRA_OEMAKE', ' VENDOR_CERT_FILE="${WORKDIR}/vendor_cert.cer"')
+
+def shim_prepare_sb_user_keys(d):
+    d.appendVar('EXTRA_OEMAKE', ' VENDOR_DBX_FILE="${WORKDIR}/vendor_dbx.esl"')
```
Comment (jiazhang0) [inline classes/user-key-store.bbclass]: You miss add VENDOR_CERT_FILE for user signing model.

File: classes/user-key-store.bbclass [hunk 8]
```diff
@@ -86,6 +86,53 @@ def mok_sb_sign(input, output, d):
...
+    cmd = (' '.join((d.getVar('STAGING_BINDIR_NATIVE', True) + '/openssl', 'pkcs12',
+           '-in', dir + 'shim_cert.pem',
+           '-inkey', dir + 'shim_cert.key',
+           '-export', '-passout', 'pass:""',
+           '-out', d.getVar('S', True) + '/shim.p12')))
+    try:
+        result, _ = bb.process.run(cmd)
+    except:
+        raise bb.build.FuncFailed('ERROR: Unable to create shim.p12')
+
```
Comment (jiazhang0) [inline classes/user-key-store.bbclass]: shim.p12 should be not used any more. This chunk should be deleted

File: classes/user-key-store.bbclass [hunk 9]
```diff
@@ -86,6 +86,53 @@ def mok_sb_sign(input, output, d):
...
+    # Prepare shim_cert and vendor_cert.
+    dir = mok_sb_keys_dir(d)
+
+    import shutil
+    shutil.copyfile(dir + 'shim_cert.pem', d.getVar('S', True) + '/shim.crt')
+    shutil.copyfile(dir + 'shim_cert.key', d.getVar('S', True) + '/shim.key')
+    pem2der(dir + 'vendor_cert.pem', d.getVar('WORKDIR', True) + '/vendor_cert.cer', d)
+    d.appendVar('EXTRA_OEMAKE', ' VENDOR_CERT_FILE="${WORKDIR}/vendor_cert.cer"')
+
```
Comment (jiazhang0) [inline classes/user-key-store.bbclass]: If we already don't depend on the Makefile to sign anything, is it still necessary copy the cert and key file to ${S}?

File: recipes-bsp/shim/shim_git.bb (hunk 4 above)
Comment (yunguowei) [inline recipes-bsp/shim/shim_git.bb]: fb.efi is not installed nor deployed, so I didn't sign this binary.

Comment (jiazhang0) [inline recipes-bsp/shim/shim_git.bb]: Sign it for the consistency plz.

File: classes/user-key-store.bbclass (hunk 9 above)
Comment (jiazhang0) [inline classes/user-key-store.bbclass]: OK shim.crt is required to be placed under ${S} for building shim_cert.h. shim.key should be dropped. vendor_cert is fine with this.

File: classes/user-key-store.bbclass (hunk 3 above)
Comment (yunguowei) [inline classes/user-key-store.bbclass]: okay, I will deal with the two types of secure boot.

**Comment by yunguowei** [comment]
V2 chagnes:

1. add MOK_SB and UEFI_SB in sb_sign.
2. remove unnecessary copy in shim_keys preparation.
3. introduce a new api, shim_sb_sign as shim could sign with EDSS or UEFI keys, can't sign with MOK_SB keys
4. install corresponding packages in templates

File: classes/user-key-store.bbclass [hunk 10]
```diff
@@ -61,10 +62,18 @@ def mok_sb_keys_dir(d):
...
-        uefi_sb_sign(input, output, d)
+    if uks_signing_model(d) == 'sample' or uks_signing_model(d) == 'user':
+        # deal with MOK_SB firstly, as MOK_SB means MOK_SB | UEFI_SB
+        # On this scenario, bootloader is verified by shim_cert.pem
+        if d.getVar('MOK_SB', True) == '1':
+            mok_sb_sign(input, output, d)
+        # UEFI_SB is defined, but MOK_SB is not defined
+        # On this scenario, shim is not used, and DB.pem is used to
+        # verify bootloader directly.
+        elif d.getVar('UEFI_SB', True) == '1':
```
Comment (jiazhang0) [inline classes/user-key-store.bbclass]: Move this condition to the beginning of this function, and place "else" here. So for edss case, do nothing if uefi secure boot is not enabled.

File: templates/feature/mok-secure-boot/template.conf [hunk 11]
```diff
@@ -4,3 +4,5 @@
 
 # Indicate the feature of Mok secure boot is enabled
 MOK_SB = "1"
+
+CUBE_ESSENTIAL_EXTRA_INSTALL_append += "packagegroup-mok-secure-boot"
```
Comment (jiazhang0) [inline templates/feature/mok-secure-boot/template.conf]: Plz don't do this. The essential stuff is overc-specific, but this layer is common.

File: templates/feature/tpm2/template.conf [hunk 12]
```diff
@@ -3,3 +3,5 @@
 #
 
 MACHINE_FEATURES += "tpm2"
+
+CUBE_ESSENTIAL_EXTRA_INSTALL_append += "packagegroup-tpm2"
```
Comment (jiazhang0) [inline templates/feature/tpm2/template.conf]: ditto

File: templates/feature/uefi-secure-boot/template.conf [hunk 13]
```diff
@@ -10,3 +10,5 @@ KERNEL_FEATURES_append_x86-64 = " cfg/efi.scc"
 
 # Always bundle the initramfs as long as initramfs is used
 INITRAMFS_IMAGE_BUNDLE = "${@'1' if d.getVar('INITRAMFS_IMAGE', True) else ''}"
+
+CUBE_ESSENTIAL_EXTRA_INSTALL_append += "packagegroup-uefi-secure-boot"
```
Comment (jiazhang0) [inline templates/feature/uefi-secure-boot/template.conf]: ditto

File: classes/user-key-store.bbclass [hunk 14]
```diff
@@ -75,6 +75,12 @@ def sb_sign(input, output, d):
     elif uks_signing_model(d) == 'edss':
         edss_sign_efi_image(input, output, d)
 
+def shim_sb_sign(input, output, d):
+    if uks_signing_model(d) == 'edss':
+        edss_sign_efi_image(input, output, d)
+    else:
+        uefi_sb_sign(input, output, d)
+
```
Comment (jiazhang0) [inline classes/user-key-store.bbclass]: Put non-edss case on top of edss case.

**Comment by yunguowei** [comment]
V3 changes:
1. drop appending package install in template.conf for cube-essential.
2. adjust judge condition, making sure signing model "user" and "sample" are processed at first.

**Comment by jiazhang0** [comment]
V3 looks good enough. TA.

**Comment by lifupan** [comment]
merged

Fupan

**Pull request merged**